test/e2e/elasticache/snapshot/e2e.sh

```diff
@@ -0,0 +1,276 @@
+  sleep 10
+  local cc_id="$rg_id-001"
+  local snapshot_yaml=$(cat <<EOF
+apiVersion: elasticache.services.k8s.aws/v1alpha1
+kind: Snapshot
+metadata:
+  name: $snapshot_name
+spec:
+  snapshotName: $snapshot_name
+  replicationGroupID: $rg_id
+  cacheClusterID: $cc_id
+EOF)
+  echo "$snapshot_yaml" | kubectl apply -f -
+  assert_equal "0" "$?" "Expected application of $snapshot_name to succeed" || exit 1
+  k8s_wait_resource_synced "snapshots/$snapshot_name" 20
+
+  # delete snapshot for case 2 if creation succeeded
+  kubectl delete snapshots/"$snapshot_name"
+  aws_wait_snapshot_deleted "$snapshot_name"
+}
+
+# test snapshot creation while specifying KMS key
+test_snapshot_create_KMS() {
+  debug_msg "executing ${FUNCNAME[0]}"
+
+  # create KMS key and get key ID
+  local output=$(daws kms create-key --output json)
+  assert_equal "0" "$?" "Expected creation of KMS key to succeed" || exit 1
+
+  local key_id=$(echo "$output" | jq -r -e ".KeyMetadata.KeyId")
+  assert_equal "0" "$?" "Key ID does not exist for KMS key" || exit 1
+
+  # delete replication group if it already exists (we want it to be created to below specification)
+  clear_rg_parameter_variables
+  rg_id="snapshot-test-kms"
+  daws elasticache describe-replication-groups --replication-group-id "$rg_id" 1>/dev/null 2>&1
+  if [[ "$?" == "0" ]]; then
+    daws elasticache delete-replication-group --replication-group-id "$rg_id" 1>/dev/null 2>&1
+    aws_wait_replication_group_deleted  "$rg_id" "FAIL: expected replication group $rg_id to have been deleted in ${service_name}"
+  fi
+
+  # create cluster mode disabled replication group for snapshot
+  num_node_groups=1
+  replicas_per_node_group=0
+  automatic_failover_enabled="false"
+  multi_az_enabled="false"
+  output_msg=$(provide_replication_group_yaml | kubectl apply -f - 2>&1)
+  exit_if_rg_config_application_failed $? "$rg_id"
+  wait_and_assert_replication_group_available_status
+
+  # create snapshot while specifying KMS key
+  local snapshot_name="snapshot-kms"
+  daws elasticache delete-snapshot --snapshot-name "$snapshot_name" 1>/dev/null 2>&1
+  sleep 10
+  local cc_id="$rg_id-001"
+  local snapshot_yaml=$(cat <<EOF
+apiVersion: elasticache.services.k8s.aws/v1alpha1
+kind: Snapshot
+metadata:
+  name: $snapshot_name
+spec:
+  snapshotName: $snapshot_name
+  cacheClusterID: $cc_id
+  kmsKeyID: $key_id
+EOF)
+  echo "$snapshot_yaml" | kubectl apply -f -
+  assert_equal "0" "$?" "Expected application of $snapshot_name to succeed" || exit 1
+  k8s_wait_resource_synced "snapshots/$snapshot_name" 20
+
+  # delete snapshot for case 1 if creation succeeded
+  kubectl delete snapshots/"$snapshot_name"
+  aws_wait_snapshot_deleted "$snapshot_name"
+}
+
+# run tests
+test_snapshot_CRUD
+test_snapshot_CMD_creates #issue: second snapshot doesn't have "status" property - problem with yaml or something else?
+test_snapshot_CME_creates #same issue as above
+test_snapshot_create_KMS #IAM role needs KMS access for this to work
```

Comment:
I know this is how some of the other test scripts are structured, but instead of wrapping the different tests in Bash functions like this, consider just adding a new Bash script/file that contains the test setup, test assertions. Makes it easier to read, IMHO.